bluesky/plans.py

```diff
@@ -2691,6 +2691,148 @@ def inner_relative_adaptive_scan():
+        if False (default), always scan from start to stop
+    md : dict, optional
+        metadata
+
+    Example
+    -------
+    motor = Mover('motor', {'motor': lambda x: x}, {'x': 0})
+    det = SynGauss('det', motor, 'motor', center=-1.3, Imax=1e5, sigma=0.05)
+    RE(tune_centroid([det], "det", motor, -1.5, -0.5, 0.01, 10))
+    """
+    if min_step <= 0:
+        raise ValueError("min_step must be positive")
+    if step_factor <= 0:
+        raise ValueError("step_factor must be positive")
+    if (num_points - 1) <= 2*step_factor:
+        raise ValueError(
+            "Increase num_points and/or decrease step_factor"
+            " or tune_centroid will never converge to a solution"
+        )
+    _md = {'detectors': [det.name for det in detectors],
+           'motors': [motor.name],
+           'plan_args': {'detectors': list(map(repr, detectors)),
+                         'motor': repr(motor),
+                         'start': start,
+                         'stop': stop,
+                         'num_points': num_points,
+                         'min_step': min_step,},
+           'plan_name': 'tune_centroid',
+           'hints': {},
+          }
+    _md.update(md or {})
+    try:
+        dimensions = [(motor.hints['fields'], 'primary')]
+    except (AttributeError, KeyError):
+        pass
+    else:
+        _md['hints'].setdefault('dimensions', dimensions)
+
+    @stage_decorator(list(detectors) + [motor])
+    @run_decorator(md=_md)
+    def _tune_core(start, stop, num_points, signal):
+        next_pos = start
+        step = (stop - start) / (num_points - 1)
+        peak_position = None
+        cur_I = None
+        cur_det = {}
+        sum_I = 0       # for peak centroid calculation, I(x)
+        sum_xI = 0
+        
+        while abs(step) >= min_step:
+            yield Msg('checkpoint')
+            yield from mv(motor, next_pos)
+            yield Msg('create', None, name='primary')
+            for det in detectors:
+                yield Msg('trigger', det, group='B')
+            yield Msg('wait', None, 'B')
+            for det in separate_devices(detectors + [motor]):
+                cur_det = yield Msg('read', det)
+                if signal in cur_det:
+                    cur_I = cur_det[signal]['value']
+                    sum_I += cur_I
+                    position = motor.read()[motor.name]["value"]
+                    sum_xI += position * cur_I
+
+            yield Msg('save')
+
+            if (stop - start) < abs(stop - start):
+                in_range = start >= next_pos >= stop  # negative motion
+            else:
+                in_range = start <= next_pos <= stop  # positive motion
+            
+            if in_range:
+                next_pos += step
+            else:
+                if sum_I == 0:
+                    return
+                peak_position = sum_xI / sum_I  # centroid
+                # TODO: report current peak_position in metadata
```

Comment:
Indeed, I don't think we have a very good story about how to get results like this out of a plan. We need one. Perhaps we can think about that in conjunction with #862.

Reply:
Right.  This case is easy since it sets the motor to the centroid, just read the motor position after RE finishes.  In general, we need a way to hook this up.  I'd want to be able to insert this content into a document.  Did not see an easy (or slightly complicated) example in the existing code.

Reply:
A complicated way to get this into the document stream now

```python
from ophyd.signal import Signal
sig = Signal(name='syn_signal', value=0)
...
yield from bp.mv(sig, your_value)
yield from trigger_and_read([sig], name='computed_results')
```

We do not currently have hooks into the `RunEngine` to pass a message with a document payload, but maybe that is something we should add.

When we started building this system out we had a number of long (and veering towards extensional) discussions about 'raw' data vs 'computed data' and what should be in the data coming out of the `RunEngine`.  We ended up settling on "data that you can read from hardware".   These were before we had the concept of streams inside of a run as pinned down as we do now and we have subsequently chipped away at this with `AttributeSignal` and `DerivedSignal` on ophyd objects.  

We probably should re-address this discussion in the near future, but I do not think it needs to be done for 1.0.